src/ServiceControl.Persistence.RavenDb/Throughput/ThroughputDataStore.cs

```diff
@@ -0,0 +1,287 @@
+            cancellationToken);
+
+        var endpoint = document?.ToEndpoint();
+        if (endpoint != null)
+        {
+            var timeSeries = await session
+                .IncrementalTimeSeriesFor(documentId, ThroughputTimeSeriesName)
+                .GetAsync(token: cancellationToken);
+
+            endpoint.LastCollectedDate = DateOnly.FromDateTime(timeSeries.LastOrDefault()?.Timestamp ?? DateTime.MinValue);
+        }
+
+        return endpoint;
+    }
+
+    public async Task<IEnumerable<(EndpointIdentifier, Endpoint?)>> GetEndpoints(IEnumerable<EndpointIdentifier> endpointIds, CancellationToken cancellationToken)
+    {
+        var endpoints = new List<(EndpointIdentifier, Endpoint?)>();
+
+        var store = await storeProvider.GetDocumentStore(cancellationToken);
+        using IAsyncDocumentSession session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        var documentIdLookup = endpointIds.ToDictionary(
+            endpointId => endpointId,
+            endpointId => endpointId.GenerateDocumentId());
+
+        var endpointDocuments = await session.LoadAsync<EndpointDocument>(
+            documentIdLookup.Values.Distinct(),
+            builder => builder.IncludeTimeSeries(ThroughputTimeSeriesName),
+            cancellationToken);
+
+        foreach (var (documentId, endpointDocument) in endpointDocuments)
+        {
+            var id = endpointDocument == null
+             ? endpointIds.First(id => id.GenerateDocumentId().Equals(documentId))
+             : endpointDocument.EndpointId;
+
+            var endpoint = endpointDocument?.ToEndpoint();
+            if (endpoint != null)
+            {
+                var timeSeries = await session
+                    .IncrementalTimeSeriesFor(documentId, ThroughputTimeSeriesName)
+                    .GetAsync(token: cancellationToken);
+
+                endpoint.LastCollectedDate = DateOnly.FromDateTime(timeSeries.LastOrDefault()?.Timestamp ?? DateTime.MinValue);
+            }
+
+            endpoints.Add((id, endpoint));
+        }
+
+        return endpoints;
+    }
+
+    public async Task SaveEndpoint(Endpoint endpoint, CancellationToken cancellationToken)
+    {
+        var document = endpoint.ToEndpointDocument();
+
+        var store = await storeProvider.GetDocumentStore(cancellationToken);
+        using IAsyncDocumentSession session = store.OpenAsyncSession(databaseConfiguration.Name);
+
+        await session.StoreAsync(document, document.GenerateDocumentId(), cancellationToken);
+        await session.SaveChangesAsync(cancellationToken);
+    }
+
+    public async Task<IDictionary<string, IEnumerable<ThroughputData>>> GetEndpointThroughputByQueueName(IEnumerable<string> queueNames, CancellationToken cancellationToken)
+    {
+        var results = queueNames.ToDictionary(queueName => queueName, queueNames => new List<ThroughputData>() as IEnumerable<ThroughputData>);
```

Comment:
Possible multiple enumeration of `queueNames` ?

Reply:
Fixed